src/ripple/app/tx/impl/XChainBridge.cpp

```diff
@@ -0,0 +1,2009 @@
+   There's a bootstrap problem. To address this, there is a special transaction
+   used to create accounts. This transaction does not require a claim id.
+
+   See the document "docs/bridge/spec.md" for a full description of how
+   bridges and their transactions work.
+*/
+
+namespace {
+
+enum class CanCreateDstPolicy { no, yes };
+
+enum class DepositAuthPolicy { normal, dstCanBypass };
+
+/** Transfer funds from the src account to the dst account
+
+    @param psb The payment sandbox.
+    @param src The source of funds.
+    @param dst The destination for funds.
+    @param dstTag Integer destination tag. Used to check if funds should be
+           transferred to an account with a `RequireDstTag` flag set.
+    @param claimOwner Owner of the claim ledger object.
+    @param amt Amount to transfer from the src account to the dst account.
+    @param canCreate Flag to determine if accounts may be created using this
+           transfer.
+    @param depositAuthPolicy Flag to determine if dst can bypass deposit auth if
+           it is also the claim owner.
+    @param j Log
+
+    @return tesSUCCESS if payment succeeds, otherwise the error code for the
+            failure reason.
+ */
+
+TER
+transferHelper(
+    PaymentSandbox& psb,
+    AccountID const& src,
+    AccountID const& dst,
+    std::optional<std::uint32_t> const& dstTag,
+    std::optional<AccountID> const& claimOwner,
+    STAmount const& amt,
+    CanCreateDstPolicy canCreate,
+    DepositAuthPolicy depositAuthPolicy,
+    beast::Journal j)
+{
+    if (dst == src)
+        return tesSUCCESS;
+
+    auto const dstK = keylet::account(dst);
+    if (auto sleDst = psb.read(dstK))
+    {
+        // Check dst tag and deposit auth
+
+        if ((sleDst->getFlags() & lsfRequireDestTag) && !dstTag)
+            return tecDST_TAG_NEEDED;
+
+        // If the destination is the claim owner, and this is a claim
+        // transaction, that's the dst account sending funds to itself. It
+        // can bypass deposit auth.
+        bool const canBypassDepositAuth = dst == claimOwner &&
+            depositAuthPolicy == DepositAuthPolicy::dstCanBypass;
+
+        if (!canBypassDepositAuth && (sleDst->getFlags() & lsfDepositAuth) &&
+            !psb.exists(keylet::depositPreauth(dst, src)))
+        {
+            return tecNO_PERMISSION;
+        }
+    }
+    else if (!amt.native() || canCreate == CanCreateDstPolicy::no)
+    {
+        return tecNO_DST;
+    }
+
+    if (amt.native())
+    {
+        auto const sleSrc = psb.peek(keylet::account(src));
+        assert(sleSrc);
+        if (!sleSrc)
+            return tecINTERNAL;
+
+        {
+            auto const ownerCount = sleSrc->getFieldU32(sfOwnerCount);
+            auto const reserve = psb.fees().accountReserve(ownerCount);
+
+            if ((*sleSrc)[sfBalance] < amt + reserve)
+            {
+                return tecINSUFFICIENT_FUNDS;
+            }
```

Comment:
There's a sneaky little bug here, or at least an inconsistency.  All of the other payment methods I'm aware of (payment, payment channel, check cash) allow the _fee_ for the transaction to chew into the reserve of an account.  But as it stands right now, an `XChainCommit` transaction does not allow that.

The reason for that difference in behavior is that `XChainCommit` is checking the `Balance` field on the account root.  In `doApply` the fee has already been removed from the account root.  So an `XChainCommit` that takes an account down to its current reserve will be rejected with `tecINSUFFICIENT_FUNDS`.

In contrast, a `Payment` does not compare to the account root `Balance` field.  It compares to the member variable `mPriorBalance`.  That member holds the account balance before the fee was removed.  So a `Payment` can take an account exactly down to its reserve and chew into the reserve for the fee.

I think this ought to be fixed, just for consistency with the other payment-like things.

Let me know if you have questions.

Oh, and while you're in there, Payment, PayChan, and CheckCash all return `tecUNFUNDED_PAYMENT`, not `tecINSUFFICIENT_FUNDS`.  You could consider also changing the return code here to `tecUNFUNDED_PAYMENT ` for better consistency.

Reply:
Fixed in 2b425a1461 `[fold] Allow fee to dip into reserve`. 

I'll added the change, but I'll note that I prefer that cross-chain transactions don't support this. Allowing regular payment fees to dip into reserves makes some sense, but extending that to cross chain transactions complicates the code to support a use-case that isn't really relevant to cross-chain transactions. However, I understand the argument for being consistent with payments, and made the change.